blockstack/gaia
# Implement `maxUploadFileSize`

Progress towards [Production ready node.js HTTP server ](https://github.com/blockstack/gaia/issues/253)

Completes issues:
* https://github.com/blockstack/gaia/issues/161 GET /hub_info should advertise the hub's maximum file size 
* https://github.com/blockstack/gaia/issues/187 Edit gaia hub drivers to check content-size on gaia hub writes 


When a content-length header is provided (almost always the case) then it is checked against the hub's max configured size. If the size is too large, the http request is immediately rejected with a specific error for the client (blockstack.js) to be able to handle appropriately. 


Additionally, while the a request payload is being streamed, the size is monitored and if the content-length is exceeded, the stream is immediately destroyed. 


If a content-length header is not provided, then the max hub size limit is used instead. This is to allow for clients that do not necessarily know the content-length in advance. The `fetch` w3c spec specifies that a ReadableStream (which has an unknown length) should be allowed as the POST body payload. This is a feature we may want to use in the future for streaming encryption and network request. 


## Manual Testing

TODO: Memory load testing during large file writes to ensure there are no performance regressions. 

## Automated Testing

Near full test coverage. SuperAgent is a PITA and difficult to spoof malicious client behavior. Some error paths are not covered in the http code reporting layer, but covered in the error handling layer right beneath it. 

## Submission Checklist


      contains correct TypeFlow annotations.


      would be expected to interact with them.
